### ISSUE TITLE
Install and Update Addons In Host DB

### ISSUE THREAD

**Issue by lasley** [Tecnativa/docker-odoo-base#31]
The next feature I am looking to add here is the automatic installation and update of addons that are defined in the `addons.yml`. 

Only reason I'm making this issue TBH is that I can't figure out exactly where you are creating the Odoo DB. 

My searches lead me [here](https://github.com/Tecnativa/docker-odoo-base/blob/a440825352b52bd3ad0f1d320fb33d2c2d127e6a/bin/pot#L10), but I'm pretty sure I'm totally off. Would you mind enlightening me please so that I can submit a shiny new PR?

**Comment by Yajo** [comment]
Neat feature!

The script you link is just something that outputs the addon's translation template `.pot` file. A quick way to get it, so to speak. [Check the docs](https://github.com/Tecnativa/docker-odoo-base#pot).

The database is created at run time, the only time when we have a ready `db` container. In fact, you can check [this entrypoint script that ensures Odoo boots after postgres is listening](https://github.com/Tecnativa/docker-odoo-base/blob/master/entrypoint.d/20-postgres-wait); and remember entrypoints get executed only at run time.

So, what would I do if I were you?

1. Add a new function to `odoobaselib` that detects the full list of used addons. Remember the little handicap here: we don't list in that file those of core odoo nor the private ones.
2. Add a new script (i.e. `bin/addoninstall`) that autoinstalls all of them (or updates them). It should end up running something like `odoo --stop-after-init -i addon1,addon2 -u addon1,addon2`, although I'm not completely sure if using `-i` and `-u` at the same time will have the desired effect.
3. Add a new `ENV ADDON_AUTOINSTALL=false` to Dockerfile.
3. Add an `entrypoint.d/40-addoninstall` that runs the script created in step 2 in case an env variable is set.

This way you can do it automatically at entrypoint time, or while your current Odoo is running with simply `docker-compose run --rm odoo addoninstall`.

Keep in mind that this entrypoint will mean **a very long downtime** when rebuilding. But if you keep current default behavior, I'm not opposed to adding it.

Odoo 9 and 10 create the database automatically on first run.

**Comment by lasley** [comment]
Thanks for the pointers, @Yajo - perfect!

Regarding the downtime - my plan is actually to use two instances for this. An update would surpass the threshold for my health checks, so it is required to have a face-forward instance and an updating one.

My hope is that I can simply leave the other Odoo instance live & upgrade the modules from underneath it. That's going to generate errors though, so I think the solution is to:
* Make the current instance read-only
* Fire up new container using upgrade entrypoint and a clone of the postgres db
* Once upgraded instance is ready, switch over to it
* Kill old instance and postgres containers

Crossing my fingers for a zero downtime upgrade here 🤞

**Comment by Yajo** [comment]
FYI current support on that is as simple as (for `prod.yaml` env):

```bash
cd /path/to/project
git pull
docker-compose -f prod.yaml build --pull  # Updates your image
docker-compose -f prod.yaml run --rm odoo odoo --stop-after-init -u base  # Updates addons with new image
docker-compose -f prod.yaml up -d  # Get the updated code facing frontend. ~ 2s of downtime
```

**Comment by pedrobaeza** [comment]
The only drawback with this method is concurrent access errors, but you just need to retry the process.

**Comment by lasley** [comment]
Wouldn't there also be a possibility of the models shifting out from underneath the live Odoo instance?

**Comment by lasley** [comment]
Now that I'm thinking about it, I think we'll only need to worry about installation of modules. Once we start talking updating, we can leverage something like OCA/server-tools#790

Regarding the base modules, what if we just added a `base` section in the `addons.yml`? We could just ignore that section in the symlinking, and only use it for the installations.

**Comment by Yajo** [comment]
👍 to both of your last points, although instead of `base`, we should call it `odoo/addons` (because that's where base addons will be).

Keep in mind that would change too the odoo.conf generation, and that there's still the `private` part too.

**Comment by lasley** [comment]
Closing to track in PR #41
